src/common/lib/client/realtimepresence.ts

```diff
@@ -224,92 +201,72 @@ class RealtimePresence extends EventEmitter {
       presence.clientId = clientId;
     }
 
-    switch (channel.state) {
-      case 'attached':
-        channel.sendPresence(presence, callback);
-        break;
-      case 'attaching':
-        this.pendingPresence.push({
-          presence: presence,
-          callback: callback,
-        });
-        break;
-      case 'initialized':
-      case 'failed': {
-        /* we're not attached; therefore we let any entered status
-         * timeout by itself instead of attaching just in order to leave */
-        const err = new PartialErrorInfo('Unable to leave presence channel (incompatible state)', 90001);
-        callback?.(err);
-        break;
+    return new Promise((resolve, reject) => {
+      switch (channel.state) {
+        case 'attached':
+          channel.sendPresence(presence, (err) => (err ? reject(err) : resolve()));
+          break;
+        case 'attaching':
+          this.pendingPresence.push({
+            presence: presence,
+            callback: (err) => (err ? reject(err) : resolve()),
+          });
+          break;
+        case 'initialized':
+        case 'failed': {
+          /* we're not attached; therefore we let any entered status
+           * timeout by itself instead of attaching just in order to leave */
+          throw new PartialErrorInfo('Unable to leave presence channel (incompatible state)', 90001);
```

Comment:
should call reject with this error

Reply:
Do you have an example of how this is inconsistent with the rest of the codebase? I can't, off the top of my head, think of other places where we wish to reject a promise directly inside the `Promise` constructor’s executor function (i.e. not as the result of an asynchronous operation). `throw`-ing an error seems like the most natural way to handle it.

Reply:
> Do you have an example of how this is inconsistent with the rest of the codebase?

You're right, there are not so many of them. Actually, there is only [one place](https://github.com/ably/ably-js/pull/1541/files#diff-89e4e931a4a9fa77b03349795008725834fd3898ee8fdd31cc83509265929e64R238) that does it, and it's a part of this PR too.

When reviewing the PR the first time I thought these places in `realtimechannel` do that too: [1](https://github.com/ably/ably-js/pull/1541/files#diff-94e592083882d1fcef5156d5f972441c3932403b3f1fce9588802dac216e9142R386), [2](https://github.com/ably/ably-js/pull/1541/files#diff-94e592083882d1fcef5156d5f972441c3932403b3f1fce9588802dac216e9142R393) (since the code around there looks so similar to the one in question in `realtimepresence`), but no, they are rejecting inside the `.once` callback.

However, as a matter of fact, I also couldn't find any other place that is `throw`-ing an error directly inside the Promise constructor. Understandable, since Promises weren't used that much before.

So I guess that'a first and we can agree on a preferred style here going forward.

> `throw-ing` an error ... requires fewer changes upon later removing the use of the Promise constructor

True, but if we were to change the code inside the Promise in the opposite direction: add callback/asynchronous operation (for example wrap this code also in `.once()`, just like in `realtimechannel`), and we forgot to change `throw` statement to `reject` call, we would get uncaught promise errors now - which is much worse than needing to change a couple of lines when removing Promise constructor and changing `reject` to `throw`.

What do you think?

Reply:
Thanks for pointing out that other place; I've changed it to also `throw`, too.

I see your point, but I think that we can live with that risk. I don't have strong opinions, though, happy to change them both to use `reject` if you prefer; let me know.

Reply:
I would prefer we always use `reject` inside Promise constructors.

Reply:
Sure; have changed all three to use `reject`.